packages/lodestar/src/chain/chain.ts

```diff
@@ -193,6 +194,70 @@ export class BeaconChain extends (EventEmitter as { new(): ChainEventEmitter })
     };
   }
 
+  /**
+   * Restore state cache and forkchoice from last finalized state.
+   */
+  private async restoreFromFinalized(): Promise<void> {
+    const start = Date.now();
+    const lastFinalizedState: TreeBacked<BeaconState> = await this.db.stateArchive.lastValue();
+    this.db.stateCache.add(lastFinalizedState);
+    // the block respective to finalized epoch is still in block db
+    const allBlocks = await this.db.block.values();
+    if (!allBlocks || allBlocks.length === 0) {
+      return;
+    }
+    this.logger.info(`Found ${allBlocks.length} blocks in database`);
+    if (allBlocks.length === 1) {
+      // start from scratch
+      const block = allBlocks[0];
+      const blockHash = this.config.types.BeaconBlock.hashTreeRoot(block.message);
+      if (this.config.types.Root.equals(blockHash, this.forkChoice.head().blockRoot)) {
+        this.logger.info("Chain is up to date, no need to restore");
+        return;
+      }
+    }
+    const finalizedStateRoot = this.config.types.BeaconState.hashTreeRoot(lastFinalizedState);
+    const sortedBlocks = allBlocks.sort((a, b) => a.message.slot - b.message.slot);
+    const isCheckpointNotGenesis = lastFinalizedState.slot > GENESIS_SLOT;
+    const finalizedBlock = sortedBlocks.find(block => {
+      return (block.message.slot === lastFinalizedState.slot) &&
+      // at genesis the genesis block's state root is not equal to genesis state root
+      (isCheckpointNotGenesis? this.config.types.Root.equals(block.message.stateRoot, finalizedStateRoot) : true);
+    });
+    if (!finalizedBlock) {
+      this.logger.error(`Cannot find block for finalized state at slot ${lastFinalizedState.slot}`);
+    } else {
+      this.logger.info(`Found finalized block at slot ${finalizedBlock.message.slot}`);
+    }
+    const blockCheckpoint = {
+      root: this.config.types.BeaconBlock.hashTreeRoot(finalizedBlock.message),
+      epoch: computeEpochAtSlot(this.config, finalizedBlock.message.slot)
+    };
+    this.forkChoice.addBlock({
+      slot: finalizedBlock.message.slot,
+      blockRoot: this.config.types.BeaconBlock.hashTreeRoot(finalizedBlock.message),
+      stateRoot: finalizedBlock.message.stateRoot.valueOf() as Uint8Array,
+      parentRoot: finalizedBlock.message.parentRoot.valueOf() as Uint8Array,
+      justifiedCheckpoint: (isCheckpointNotGenesis)? lastFinalizedState.currentJustifiedCheckpoint : blockCheckpoint,
+      finalizedCheckpoint: (isCheckpointNotGenesis)? lastFinalizedState.finalizedCheckpoint : blockCheckpoint
+    });
+
+    for (const block of sortedBlocks) {
+      await this.receiveBlock(block, true);
+    }
+    await new Promise((resolve) => {
+      const lastBlock = sortedBlocks[sortedBlocks.length - 1];
+      this.on("processedBlock", (signedBlock) => {
+        if (signedBlock.message.slot === lastBlock.message.slot &&
+          this.config.types.SignedBeaconBlock.equals(signedBlock, lastBlock)) {
+          resolve();
+        }
+      });
+    });
```

Comment:
This feels like it would be a cool method to expose from chain as `public waitForBlockProcessed`